doubly_linked_list/doubly_linked_list.py

```diff
@@ -44,25 +44,102 @@ def __len__(self):
     return self.length
 
   def add_to_head(self, value):
-    pass
-
+    self.length +=1
+    new_node = ListNode(value = value)
+    new_node.next = None
+    new_node.prev = None
+    if self.head is None:
+      self.head = new_node
+      self.tail = new_node
+    else:
+      self.head.prev = new_node
+      new_node.next = self.head
+      self.head = new_node
+    
   def remove_from_head(self):
-    pass
+    if self.head == None:
+      return
+    else:
+      self.length -=1
+      if self.head != self.tail:
+        self.head = self.head.next
+        self.head.prev = None
+      else:
+        self.head = self.tail=None
 
   def add_to_tail(self, value):
-    pass
+    self.length +=1
+    new_node = ListNode(value=value)
+    new_node.next = None
+    new_node.prev = None
+    if self.tail is None:
+      self.head = new_node
+      self.tail = new_node
+    else:
+      self.tail.next = new_node
+      new_node.prev = self.tail
+      self.tail = new_node
 
+   
   def remove_from_tail(self):
-    pass
+    if self.head is None:
+      return
+    if self.head.next is None:
+      self.head = None
+      return
+    n = self.head
+    while n.next is not None:
+      n = n.next
+    n.prev.next = None
+
+    
 
   def move_to_front(self, node):
-    pass
+    # get to the front node (there won't be a previous node)
+    curr_node = self
+    while curr_node.prev is not None:
+      print("moving to the prev node")
+      curr_node = curr_node.prev
+    # then add the node to its head
+    curr_node.add_to_head(node)
+   
 
   def move_to_end(self, node):
-    pass
+    curr_node = self
+    while curr_node.next is not None:
+      print("moving to the next node")
+      curr_node = curr_node.next
+    curr_node.add_to_tail(node)
+    
 
   def delete(self, node):
-    pass
+    if node.next is None and node.prev is None:
+      print("deleting the only node left")
+      self.head = None
+      self.tail = None
+      self.length = 0
+    elif node.next is None:
+      print("deleting the tail node")
+      self.tail = node.prev
+      self.length -= 1
+    elif node.prev is None:
+      print("deleting the head node")
+      self.head = node.next
+      self.length -= 1
+    else:
+      print("deleting a middle node")
+      node.prev.next = node.next
+      node.next.prev = node.prev
+
```

Comment:
and your delete was giving you issues:

self.length -= 1
if not self.head and not self.tail:
____return
if self.head == self.tail:
____self.head = None
____self.tail = None
elif self.head == node:
____self.head = node.next
____node.delete()
elif self.tail == node:
____self.tail = node.prev
____node.delete()
else:
____node.delete()